### PR TITLE
Pubsub server

### PR DESCRIPTION
**Pull request by thinkingfish**
This is a pubsub implementation that's compatible with Redis' pub/sub, but it's a standalone server, not mixed with other redis features.

Also, the publisher does not need to wait for the message fanout to complete before receiving a response, which fully decouples publishing from subscribing.

This is a prototype and many error/corner cases are not yet handled. Nor are unit tests added for the corresponding modules.

### CODE REVIEW

File: src/storage/pubsub/listener.c [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+        prev = l, l = SLIST_NEXT(l, l_sle)) {
+        if (l->s == s) {
+            break;
+        }
+    }
+
+    if (prev == NULL) {
+        SLIST_REMOVE_HEAD(bucket, l_sle);
+    } else {
+        SLIST_REMOVE_AFTER(prev, l_sle);
```
Comment (seppo0010) [inline src/storage/pubsub/listener.c]: what would happen if `s` was not on the list? `prev`'d be the last element and it'll segfault or something? should there be an `ASSERT(l != NULL)`?

File: src/storage/pubsub/topic.c [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+    if (t->idx == NULL) {
+        cc_free(t);
+        return NULL;
+    }
+
+    topic_reset(t);
+    t->name.len = name->len;
+    t->name.data = cc_alloc(t->name.len);
+    if (t->name.data == NULL) {
+        cc_free(t);
```
Comment (seppo0010) [inline src/storage/pubsub/topic.c]: missing `cc_free(t->idx);`

File: src/storage/pubsub/topic.h [hunk 3]
```diff
@@ -0,0 +1,33 @@
+#pragma once
+
+#include "index.h"
+
+#include <cc_bstring.h>
+
+/*
+ * a topic is an endpoint that clients can subscribe to, equivalent to
+ * "channel" in the original redis protocol.
```
Comment (seppo0010) [inline src/storage/pubsub/topic.h]: is there any reason to call it topic instead of channel? do we refer to something else as channel? you could add the reason to the comment here

File: src/server/pubsub/admin/process.c [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+void
+admin_process_teardown(void)
+{
+    log_info("tear down the %s module", PUBSUB_ADMIN_MODULE_NAME);
+    if (!admin_init) {
+        log_warn("%s has never been setup", PUBSUB_ADMIN_MODULE_NAME);
+    }
+
```
Comment (seppo0010) [inline src/server/pubsub/admin/process.c]: `cc_free(buf);`?

File: src/storage/pubsub/listener.h [hunk 5]
```diff
@@ -0,0 +1,32 @@
+#pragma once
+
+#include "index.h"
+
+#include <stream/cc_sockio.h>
+
+/*
+ * a listener is a client that has subscribed to at least one channel
```
Comment (seppo0010) [inline src/storage/pubsub/listener.h]: I thought that a listener may be subscribed to 0 channels (after unsubscribe it is not deleted)?

Comment (thinkingfish) [inline src/storage/pubsub/listener.h]: The connection is still around, but no longer a listener. This is explained on redis.io in the [pubsub section](https://redis.io/topics/pubsub):

> unsubscribe: means that we successfully unsubscribed from the channel given as second element in the reply. The third argument represents the number of channels we are currently subscribed to. When the last argument is zero, we are no longer subscribed to any channel, and the client can issue any kind of Redis command as we are outside the Pub/Sub state.

File: src/storage/pubsub/topic.h (hunk 3 above)
Comment (thinkingfish) [inline src/storage/pubsub/topic.h]: Oh yea I should explain. mostly ccommon (which Pelikan depends on and includes) already have the concept of a "channel" to mean some kind of connection (tcp, uds, pipe...)

File: src/core/data/pubsub.c [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+
+        event_del(ctx->evb, hdl->wid(c));
+        event_add_write(ctx->evb, hdl->wid(c), s);
+    } else if (status == CC_ERROR) {
+        c->state = CHANNEL_TERM;
+    }
+
+    if (processor->write(s) < 0) {
+        log_debug("handler signals channel termination");
+        s->ch->state = CHANNEL_TERM;
```
Comment (thinkingfish) [inline src/core/data/pubsub.c]: done

File: src/server/pubsub/data/process.c [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+ * reply: # listeners subscribed
+ */
+static void
+command_publish(struct response *rsp, struct request *req, struct buf_sock *s)
+{
+    struct element *el_t, *el_m; /* topic & message */
+    struct element el_r = {.type = ELEM_INT}; /* reply, an integer */
+    struct listener *l;
+    struct topic *t;
+    uint32_t nsub = 0;
```
Comment (thinkingfish) [inline src/server/pubsub/data/process.c]: you are right, I changed it.

File: src/server/pubsub/data/process.c [hunk 8]
```diff
@@ -0,0 +1,337 @@
...
+            /* handle error */
+            topic_put(t);
+        }
+
+        log_verb("subscribing to topic %.*s", el->bstr.len, el->bstr.data);
+        if (!topic_add_listener(t, l)) {
+            log_debug("listener not added");
+        }
+        if (!listener_add_topic(l, t)) {
+            log_debug("topic not added");
```
Comment (thinkingfish) [inline src/server/pubsub/data/process.c]: Let me keep you comment in mind for the moment- I need to handle errors in a number of other places as well but it's not immediately clear what Redis does/should do in this case. I'll research more and come back to this.

File: src/storage/pubsub/listener.c (hunk 1 above)
Comment (thinkingfish) [inline src/storage/pubsub/listener.c]: You are right, the way I'm using it now assumes `l != NULL`, but I might as well allow passing in an invalid listener to simplify the deletion logic.

Eventually I want to add pooling to listener (and topic) to avoid calling malloc & free all the time, I will later think about if I want to allow people to create/destroy listener objects directly.

File: src/storage/pubsub/topic.c (hunk 2 above)
Comment (thinkingfish) [inline src/storage/pubsub/topic.c]: fixed

File: src/server/pubsub/admin/process.c (hunk 4 above)
Comment (thinkingfish) [inline src/server/pubsub/admin/process.c]: fixed

File: src/core/core.c [hunk 9]
```diff
@@ -10,7 +10,7 @@
 #include <sysexits.h>
 
 void
-core_run(void *arg_worker)
+worker_run(void *arg_worker)
```
[the remark this answers is not in the document]
Comment (kevyang) [reply src/core/core.c]: maybe rename these to core_worker_run and core_pubsub_run
